acceptance-tests/https_frontend_test.go

```diff
@@ -41,40 +52,87 @@ var _ = Describe("HTTPS Frontend", func() {
       alternative_names: [haproxy.internal]
 `
 
-		haproxyBackendPort := 12000
-		haproxyInfo, varsStoreReader := deployHAProxy(baseManifestVars{
+	var creds struct {
+		HTTPSFrontend struct {
+			Certificate string `yaml:"certificate"`
+			PrivateKey  string `yaml:"private_key"`
+			CA          string `yaml:"ca"`
+		} `yaml:"https_frontend"`
+	}
+
+	JustBeforeEach(func() {
+		var varsStoreReader varsStoreReader
+		haproxyInfo, varsStoreReader = deployHAProxy(baseManifestVars{
 			haproxyBackendPort:    haproxyBackendPort,
 			haproxyBackendServers: []string{"127.0.0.1"},
 			deploymentName:        defaultDeploymentName,
-		}, []string{opsfileSSLCertificate}, map[string]interface{}{}, true)
-
-		var creds struct {
-			HTTPSFrontend struct {
-				Certificate string `yaml:"certificate"`
-				PrivateKey  string `yaml:"private_key"`
-				CA          string `yaml:"ca"`
-			} `yaml:"https_frontend"`
-		}
+		}, []string{opsfileHTTPS}, map[string]interface{}{
+			"enable_http2": enableHTTP2,
+		}, true)
 
 		err := varsStoreReader(&creds)
 		Expect(err).NotTo(HaveOccurred())
 
-		closeLocalServer, localPort := startDefaultTestServer()
-		defer closeLocalServer()
-
-		closeTunnel := setupTunnelFromHaproxyToTestServer(haproxyInfo, haproxyBackendPort, localPort)
-		defer closeTunnel()
+		var localPort int
+		closeLocalServer, localPort = startDefaultTestServer()
+		closeTunnel = setupTunnelFromHaproxyToTestServer(haproxyInfo, haproxyBackendPort, localPort)
 
-		client := buildHTTPClient(
+		http1Client = buildHTTPClient(
 			[]string{creds.HTTPSFrontend.CA},
 			map[string]string{"haproxy.internal:443": fmt.Sprintf("%s:443", haproxyInfo.PublicIP)},
 			[]tls.Certificate{},
 		)
+	})
+
+	AfterEach(func() {
+		if closeLocalServer != nil {
+			defer closeLocalServer()
+		}
+		if closeTunnel != nil {
+			defer closeTunnel()
+		}
+	})
 
-		By("Sending a request to HAProxy")
-		resp, err := client.Get("https://haproxy.internal:443")
+	It("Correctly proxies HTTPS requests", func() {
+		By("Sending a request to HAProxy using HTTP 1.1")
+		resp, err := http1Client.Get("https://haproxy.internal:443")
 		Expect(err).NotTo(HaveOccurred())
+
+		Expect(resp.ProtoMajor).To(Equal(1))
+
 		Expect(resp.StatusCode).To(Equal(http.StatusOK))
 		Eventually(gbytes.BufferReader(resp.Body)).Should(gbytes.Say("Hello cloud foundry"))
 	})
+
+	Context("When ha_proxy.enable_http2 is true", func() {
```

Comment:
There is one test missing:
"When ha_proxy.enable_http2 is false"
"It responds with a 400 bad request response to a H2 request"

Reply:
Hi @domdom82 it won't be a 400 because this happens at a lower level, but I'll add a test that verifies that an HTTP2-compatible client negotiates an HTTP 1.1 protocol when `ha_proxy.enable_http2` is false

Reply:
interesting. I thought of that initially but then I got a 400 when using openssl to test this. I noticed that curl always offers http/1.1 even if you put `--http2` so it sent a http/1.1 request when HAproxy only offered that. I think the go h2 client will do the same. 
It may be an academic point, but if you "forced" a H2 request to HAproxy that only supports HTTP/1.1 I am pretty sure it would give a 400. (H2 being binary and not at all looking like a HTTP/1.1 at all to HAproxy)

Reply:
Hi @domdom82 I wasn't able to find a way to cause an error with an HTTP2-compatible client -- ie seems like HAProxy happily uses HTTP2 if the client only supports HTTP2, even if `enable_http2: false`. However this test at least verifies that clients that support HTTP2 will prefer HTTP 1.1 when `enable_http2: false` https://github.com/cloudfoundry-incubator/haproxy-boshrelease/blob/23a7546a494624d17fa12228b1cd188a510de4b8/acceptance-tests/https_frontend_test.go#L117

Reply:
Actually looks like this test is failing -- so it's using HTTP2 even without `alpn h2,http/1.1`